packages/controller-utils/src/util.ts

```diff
@@ -251,20 +251,28 @@ export async function safelyExecuteWithTimeout<Result>(
 }
 
 /**
- * Convert an address to a checksummed hexidecimal address.
+ * Convert an address to a checksummed hexadecimal address.
  *
  * @param address - The address to convert.
- * @returns A 0x-prefixed hexidecimal checksummed address, if address is valid. Otherwise original input 0x-prefixe, if address is valid. Otherwise original input 0x-prefixed.
+ * @returns The address in 0x-prefixed hexadecimal checksummed form if it is valid, or untouched otherwise.
  */
-export function toChecksumHexAddress(address: string) {
+export function toChecksumHexAddress<T>(address: T) {
```

Comment:
This relies on the return type being `string | T`. Setting it to `T` causes the following error: 
```ts
Type 'string' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to 'string'.ts(2322)
```

IMO `string | T` defeats the purpose of our using the generic param `T`. 

What do you feel about adapt the overload you suggested earlier, so that it [distinctly handles string vs. non-string input](https://www.typescriptlang.org/play?#code/PQKhAIHkDcFMCcA2B7AhgE3AZwJYHMA7VAFwFd5YtwRgAoWADwAdl5jwAzUggY2J2QFwxZAGEAFrB4BrLKQC2ACUYBBdOgpYsACgwbKWAFzZi8HATwBKYwAMADAwAkAbyynzeAL42A3PWas7Fy8-ILCYpIyckqq6ppYADwAKgB8unEGxknW4El+oBAAkvJMiLDysATEJAJCuIQk5LDUdIwsbJzcfLXhElKyCsoMavpa6aNG4NzSBMgA7gQ507ML4M604JvgFGTwBH6etLQF4IUEHAhUPKwUfODIcEhomPVEu80crODmTKTsxABPJjNGhHMrsHBYABypEQiGMBFhiFo4O+WCGtgcLjcZgs3iOBQ2EAAyu4LN8CL9iES6MdgJ0Qj0RH0ooNYhNxvFjDiPDl7E5XGSvDZwNoANQARnujxQGEstEhMLh4AAvL1IgMYsMMmMAOS6yw+cDAelJIHNXX87FC7y6tHgWbsVBafBEABGZXCwnN4F1iLhuoAdNoAEwAZhDIfldIZ3TCzI10SGI3inMyJlxVkxAp5eJF4qlDwQsvQ8shQ1V6v6SfZqd1DgALLAJRwOAB2SMATjd6lQHDsqDbYfQHAlAA4w6gx2P0J2eJO7A2eDxO22AKywOwGglgIngKGCAC0ubwFKpNIJ9OCcaECerbO1EwSplIsDSei5wngr5yL+aBelYtnjLdFGErO9WS1FMDG0P9DWNU0fV1P87UhB1kCdF0Gg9ZoRG9YFfStQVM1tYNw0jaMTVjUJbwie8oJ1RJuHQWAOHMWB0HfRjjGY1j2NLHiCBYtiCA40VJUAp45QVUCGHAujIOTRjtF4kSOPgqizQI3VVP41CqEdcBnVeVAcK9QFtKIk9SNDCMo0vaimQUzUlKff1EC4iYESRHJ3PEwsZWAmSlUQeSWRc2sYPcw1aCAA), inferring the return types for each?

Here's the diff for this change, including an error fix (see https://github.com/MetaMask/core/pull/4046#discussion_r1525044828). This passes `yarn build`, `yarn test`. 

```diff
diff --git a/packages/controller-utils/src/util.ts b/packages/controller-utils/src/util.ts
index f06cf6391..e5da29a81 100644
--- a/packages/controller-utils/src/util.ts
+++ b/packages/controller-utils/src/util.ts
@@ -250,13 +250,27 @@ export async function safelyExecuteWithTimeout<Result>(
   }
 }
 
+/**
+ * Convert an address to a checksummed hexadecimal address.
+ *
+ * @param address - The address to convert.
+ * @returns A 0x-prefixed hexadecimal checksummed address, if address is valid. Otherwise returns it 0x-prefixed.
+ */
+export function toChecksumHexAddress(address: string): Hex;
+/**
+ * Returns input unmodified.
+ * @deprecated This function should only be called with string input.
+ * @param address - The address to convert.
+ * @returns Unmodified input
+ */
+export function toChecksumHexAddress<T>(address: T): T;
 /**
  * Convert an address to a checksummed hexadecimal address.
  *
  * @param address - The address to convert.
  * @returns The address in 0x-prefixed hexadecimal checksummed form if it is valid, or untouched otherwise.
  */
-export function toChecksumHexAddress<T>(address: T) {
+export function toChecksumHexAddress(address: unknown): unknown {
   if (typeof address !== 'string') {
     // Mimic behavior of `addHexPrefix` from `ethereumjs-util` (which this
     // function was previously using) for backward compatibility.
diff --git a/packages/preferences-controller/src/PreferencesController.ts b/packages/preferences-controller/src/PreferencesController.ts
index 2bf761950..b7451b2c2 100644
--- a/packages/preferences-controller/src/PreferencesController.ts
+++ b/packages/preferences-controller/src/PreferencesController.ts
@@ -254,7 +254,9 @@ export class PreferencesController extends BaseController<
    * @param addresses - List of addresses to use to generate new identities.
    */
   addIdentities(addresses: string[]) {
-    const checksummedAddresses = addresses.map(toChecksumHexAddress);
+    const checksummedAddresses = addresses.map((address) =>
+      toChecksumHexAddress(address),
+    );
     this.update((state) => {
       const { identities } = state;
       for (const address of checksummedAddresses) {

```